cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringProgramEventPublisher.java

```diff
@@ -111,6 +112,11 @@ public class TetheringProgramEventPublisher extends AbstractRetryableScheduledSe
         cConf.getLong(Constants.Tethering.CONNECTION_INTERVAL));
   }
 
+  @VisibleForTesting
+  void setMessageFetcher(MessageFetcher messageFetcher) {
```

Comment:
instead of adding this setter, we can use a TestMessageFetcher that has a `setMessages(Iterator<Message>)` method, and pass that into the constructor. That way the MessageFetcher here remains final and we don't break encapsulation in this real class, but only in the test class.

Reply:
done